### ISSUE TITLE
Unify formula and cask API source download handling

### ISSUE THREAD

**Issue by Bo98** [Homebrew/brew#14746]
### Verification

- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### Provide a detailed description of the proposed feature

The formula way to handle downloads is fully fleshed out with progress information and the cask system is not. The stage where the download occurs is also different. It's worth consolidating the two.

### What is the motivation for the feature?

* Improved UX for cask source downloads
* More common shared code (better maintainability)

### How will the feature be relevant to at least 90% of Homebrew users?

The UX for cask source downloads is poor in that there is no progress information and it happens too regularly. Fixing that will provide a better experience.

### What alternatives to the feature have been considered?

Keep it separate and just modify the cask side for feature parity.

**Comment by Bo98** [comment]
I'm working on this right now (or rather once #14733 is merged), but it might take a day or two. Just opening this so there isn't duplicated work here.

**Comment by apainintheneck** [comment]
I wonder if there'd be some way to cache the downloaded package source instead of re-downloading it every time. The fact that we're using the git revision makes that harder though. The reason I ask is that there are some commands where we really do need the source and fetching it infrequently doesn't seem like a terrible idea though it could end up being a lot of network requests. This was brought up as an option to add to the `brew cat` command (https://github.com/Homebrew/brew/issues/14815), for example, and there are a few developer commands that could work like that too in the future.

**Comment by MikeMcQuaid** [comment]
> I wonder if there'd be some way to cache the downloaded package source instead of re-downloading it every time.

Definitely could do this. Feels like it'd be a useful thing to add.

> The fact that we're using the git revision makes that harder though.

Indeed. Also want to make sure we cleanup these old versions sufficiently (although this could be done with the automatic formula/periodic `brew cleanup` runs).

**Comment by Bo98** [comment]
Still got local changes for this I'll either finish or push somewhere this weekend or Monday.

**Comment by MikeMcQuaid** [comment]
@Bo98 Thanks! IMO it's worth opening a draft PR for this sort of thing sooner rather than later.

**Comment by Bo98** [comment]
yeah I've allocated like 12 hours of Homebrew tomorrow for ops stuff so will do that alongside.
